### PR TITLE
8319999: Refactor MetaspaceShared::use_full_module_graph()

### PR DESCRIPTION
**Pull request by iklam**
This is another step of moving CDS config management into cdsConfig.hpp:

The function `MetaspaceShared::use_full_module_graph()` is split into two:
- `CDSConfig::is_dumping_full_module_graph()`
- `CDSConfig::is_loading_full_module_graph()`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319999](https://bugs.openjdk.org/browse/JDK-8319999): Refactor MetaspaceShared::use_full_module_graph() (**Sub-task** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [f370aea4](https://git.openjdk.org/jdk/pull/16646/files/f370aea4b36ee0eacf919fee3b28ab46370963f3)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16646/head:pull/16646` \
`$ git checkout pull/16646`

Update a local copy of the PR: \
`$ git checkout pull/16646` \
`$ git pull https://git.openjdk.org/jdk.git pull/16646/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16646`

View PR using the GUI difftool: \
`$ git pr show -t 16646`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16646.diff">https://git.openjdk.org/jdk/pull/16646.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16646#issuecomment-1809510192)

### CODE REVIEW

File: src/hotspot/share/cds/cdsConfig.cpp [hunk 1]
```diff
@@ -23,10 +23,15 @@
...
 #include "precompiled.hpp"
+#include "cds/archiveHeapLoader.hpp"
 #include "cds/cdsConfig.hpp"
 #include "cds/heapShared.hpp"
+#include "classfile/classLoaderDataShared.hpp"
+#include "logging/log.hpp"
 
 bool CDSConfig::_is_dumping_dynamic_archive = false;
+bool CDSConfig::_enable_dumping_full_module_graph = true;
+bool CDSConfig::_enable_loading_full_module_graph = true;
```
Comment (dholmes-ora) [inline src/hotspot/share/cds/cdsConfig.cpp:34]: The "enable" naming isn't quite right - these fields hold the enabled state e.g. `_dumping_full_module_graph_enabled`.

File: src/hotspot/share/cds/cdsConfig.hpp [hunk 2]
```diff
@@ -30,7 +30,9 @@
 
 class CDSConfig : public AllStatic {
 #if INCLUDE_CDS
-  static bool _is_dumping_dynamic_archive;
+  static bool     _is_dumping_dynamic_archive;
```
Comment (dholmes-ora) [inline src/hotspot/share/cds/cdsConfig.hpp:33]: Nit: Extra spaces

File: src/hotspot/share/cds/filemap.cpp [hunk 3]
```diff
@@ -213,7 +213,7 @@ void FileMapHeader::populate(FileMapInfo *info, size_t core_region_alignment,
   _compressed_class_ptrs = UseCompressedClassPointers;
   _max_heap_size = MaxHeapSize;
   _use_optimized_module_handling = MetaspaceShared::use_optimized_module_handling();
-  _use_full_module_graph = MetaspaceShared::use_full_module_graph();
+  _has_full_module_graph = CDSConfig::is_dumping_full_module_graph();
```
Comment (dholmes-ora) [inline src/hotspot/share/cds/filemap.cpp:216]: `_uses_full_module_graph` would seem more appropriate than `use` or `has`

**Review by dholmes-ora — COMMENTED** [review]
Seems reasonable but some of the naming seems a little off to me.

Thanks.

File: src/hotspot/share/cds/filemap.cpp (hunk 3 above)
Comment (iklam) [reply src/hotspot/share/cds/filemap.cpp:216]: The name is consistent with other fields in filemap.hpp:

```
  bool _has_non_jar_in_classpath;     // non-jar file entry exists in classpath
  bool _has_platform_or_app_classes;  // Archive contains app classes
  bool _has_full_module_graph;        // Does this CDS archive contain the full archived module graph?
```

File: src/hotspot/share/cds/cdsConfig.hpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/cds/cdsConfig.hpp:33]: The alignment is the same as the is_dumping_xxx()/enable_dumping_xxx() functions below, so it's easier to read the "dumping" and "loading" part of the names.

File: src/hotspot/share/cds/cdsConfig.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/cds/cdsConfig.cpp:34]: I like the variable name to be the same as the function name, so it's easier to search. We also have `bool _enable_preview;` in arguments.hpp

File: src/hotspot/share/cds/cdsConfig.hpp (hunk 2 above)
Comment (dholmes-ora) [reply src/hotspot/share/cds/cdsConfig.hpp:33]: Sorry that is not a valid reason IMO. It's bad enough when people want the start of names to line up, never-mind the middle of them!

File: src/hotspot/share/cds/filemap.cpp (hunk 3 above)
Comment (dholmes-ora) [reply src/hotspot/share/cds/filemap.cpp:216]: Okay makes more sense when expressed like that - but it is not obvious that the MetaspaceShared API should be read as applying "to this CDS archive".

File: src/hotspot/share/cds/cdsConfig.cpp (hunk 1 above)
Comment (dholmes-ora) [reply src/hotspot/share/cds/cdsConfig.cpp:34]: But the context is different. `enable_xxx()` indicates that the function enables xxx. The variable that holds the enabled/disabled state of xxx should reflect in its name that it holds that state so `_xxx_enabled` Or `_has_xxx`, or `_is_xxx`, depending on what `xxx` actually is.

File: src/hotspot/share/cds/cdsConfig.hpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/cds/cdsConfig.hpp:33]: That's my personal preference. Many people align the code differently. Are there any rules in the code guide about code alignments?

File: src/hotspot/share/cds/cdsConfig.cpp (hunk 1 above)
Comment (dholmes-ora) [reply src/hotspot/share/cds/cdsConfig.cpp:34]: Or `_use_xxx` as was originally used here. I guess I just find it particularly irksome that `enable` is often used where `enabled` would be the grammatically correct choice.

File: src/hotspot/share/cds/cdsConfig.hpp (hunk 2 above)
Comment (dholmes-ora) [reply src/hotspot/share/cds/cdsConfig.hpp:33]: Nope no rules as such.

**Review by dholmes-ora — APPROVED** [review]

File: src/hotspot/share/cds/cdsConfig.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/cds/cdsConfig.cpp]: I don't think enable_xxx is grammatically wrong, when it's read as "(we should) enable xxx". We have hundreds of global flags that start with Use/Enable/Disable, like EnableDynamicAgentLoading and UseTLAB.

Anyway, the reason I didn't use `_is_loading_full_module_graph` is that I want a flag that unconditionally disables the feature, but this flag alone doesn't guarantee that the feature exists.

I changed the flag to `_loading_full_module_graph_disabled` and added comments to make it clear.

File: src/hotspot/share/cds/filemap.hpp [hunk 4]
```diff
@@ -280,7 +281,6 @@ class FileMapHeader: private CDSFileMapHeaderBase {
   void set_mapped_base_address(char* p)          { _mapped_base_address = p; }
   void set_heap_roots_offset(size_t n)           { _heap_roots_offset = n; }
   void copy_base_archive_name(const char* name);
-
```
Comment (calvinccheung) [inline src/hotspot/share/cds/filemap.hpp:283]: Line removed by accident?

File: src/hotspot/share/classfile/classLoaderDataShared.cpp [hunk 5]
```diff
@@ -182,25 +182,25 @@ void ClassLoaderDataShared::serialize(SerializeClosure* f) {
 }
 
 void ClassLoaderDataShared::clear_archived_oops() {
-  assert(UseSharedSpaces && !MetaspaceShared::use_full_module_graph(), "must be");
+  assert(UseSharedSpaces && !CDSConfig::is_loading_full_module_graph(), "must be");
```
Comment (calvinccheung) [inline src/hotspot/share/classfile/classLoaderDataShared.cpp:185]: Is the `UseSharedSpaces `needed?

File: src/hotspot/share/classfile/modules.cpp [hunk 6]
```diff
@@ -599,12 +600,12 @@ void Modules::serialize(SerializeClosure* soc) {
...
     }
     log_info(cds)("optimized module handling: %s", MetaspaceShared::use_optimized_module_handling() ? "enabled" : "disabled");
-    log_info(cds)("full module graph: %s", MetaspaceShared::use_full_module_graph() ? "enabled" : "disabled");
+    log_info(cds)("full module graph: %s", CDSConfig::is_loading_full_module_graph() ? "enabled" : "disabled");
   }
 }
 
 void Modules::define_archived_modules(Handle h_platform_loader, Handle h_system_loader, TRAPS) {
-  assert(UseSharedSpaces && MetaspaceShared::use_full_module_graph(), "must be");
+  assert(UseSharedSpaces && CDSConfig::is_loading_full_module_graph(), "must be");
```
Comment (calvinccheung) [inline src/hotspot/share/classfile/modules.cpp:608]: Is the `UseSharedSpaces` needed?

**Review by calvinccheung — COMMENTED** [review]
Spotted a few minor items.

**Review by calvinccheung — APPROVED** [review]
Updates look good.

**Comment by iklam** [comment]
Thanks @dholmes-ora and @calvinccheung for the review!
/integrate

**Pull request closed without merging**